### PR TITLE
Add support for THT Resistor Color Code Markings

### PR DESCRIPTION
**Pull request by Electro707**
Related to https://github.com/yaqwsx/PcbDraw-Lib/pull/19, this PR allows PcbDraw to, if the component it's drawing is a THT resistor, to auto-detect the resistor value and change the band's color to reflect that:
image

The resistor's values can also be set manually with the `resistor-values` argument, input given as comma delimited key-value pairs that are colon delimited, for example `resistor-values R1:1k,R2:20k`.

The colors of the bands are in the default board style. If a board style doesn't have this (or any keys), it will get copied from the default style.

Closes https://github.com/yaqwsx/PcbDraw/issues/71

TODO:
- [x] Add option to "flip" the resistor, which would just flip the colors on the bands
- [x] Perhaps there is a better argument input style for the `resistor-values`
- [x] Add tolerance parsing
- [x] The resistor band colors are stored in the board style. Should this be moved to it's own style input?
- [x] If a board style doesn't have a key, it will be copied from the default style. This might be undesired. Thoughts?
- [x] Create test boards with just resistor and create a test script to verify it all works
- [ ] Document this feature
- [ ] Wait until https://github.com/yaqwsx/PcbDraw-Lib/pull/19 is merged to update the version used in this repository

### CODE REVIEW

**Comment by Electro707** [comment]
Removing the WIP tag as the rest of what needs to be done should be up for discussion.

File: pcbdraw/pcbdraw.py [hunk 1]
```diff
@@ -581,13 +605,22 @@ def component_to_board_scale(svg):
...
         return
 
-    unique_name = f"{lib}__{name}"
+    # If the part is a THT resistor, change it's value if the parameter custom_res_color has
+    # the resistor specifing a new value
+    if lib == "Resistor_THT":
+        if tht_resistor_settings is not None:
+            if ref in tht_resistor_settings:
+                if 'override_val' in tht_resistor_settings[ref]:
+                    value = tht_resistor_settings[ref]['override_val']
```
Comment (yaqwsx) [inline pcbdraw/pcbdraw.py:621]: I would prefere to not check for the library as the users might use custom libraries. We should only check if there is a resistor value specified; if so we try to color it. If it fails we notify the user and stop.

File: pcbdraw/pcbdraw.py [hunk 2]
```diff
@@ -628,6 +661,53 @@ def component_from_library(lib, name, value, ref, pos, usedComponents, comp,
         }
         usedComponents[unique_name] = componentInfo
 
+        # If the library used is the THT resistor one, attempt to change the band colors if they exsist
+        if lib == "Resistor_THT":
```
Comment (yaqwsx) [inline pcbdraw/pcbdraw.py:665]: The same here as above.

File: pcbdraw/pcbdraw.py [hunk 3]
```diff
@@ -628,6 +661,53 @@ def component_from_library(lib, name, value, ref, pos, usedComponents, comp,
...
+                    for i in range(len(s)):
+                        if s[i].startswith('fill:'):
+                            s_split = s[i].split(':')
+                            s_split[1] = res_c
+                            s[i] = ':'.join(s_split)
+                        elif s[i].startswith('opacity:'):
+                            s_split = s[i].split(':')
+                            s_split[1] = '1'
+                            s[i] = ':'.join(s_split)
+                    band.attrib["style"] = ";".join(s)
```
Comment (yaqwsx) [inline pcbdraw/pcbdraw.py:709]: Plese consider factoring out this code into a separate function. `component_from_library` is already too long and hard to follow.

File: pcbdraw/pcbdraw.py [hunk 4]
```diff
@@ -867,6 +949,24 @@ def main():
         print(e)
         sys.exit(1)
 
+    # If some keys aren't in the loaded style compared to the default style, copy it from the default style
+    for s in default_style:
+        if s not in style:
+            style[s] = default_style[s]
```
Comment (yaqwsx) [inline pcbdraw/pcbdraw.py:955]: If we inherit properties from default styles, we should also check that there are no extra keys in the user-supplied style. When we don't check for them it is really hard to spot typos in key names (since it silently fails)

File: pcbdraw/pcbdraw.py [hunk 5]
```diff
@@ -867,6 +949,24 @@ def main():
...
+
+    tht_resistor_settings = {}
+    if args.resistor_values:
+        split_list = args.resistor_values.split(",")
+        for r in split_list:
+            r_s = r.split(":")
+            tht_resistor_settings[r_s[0]] = {'override_val': r_s[1]}
```
Comment (yaqwsx) [inline pcbdraw/pcbdraw.py:962]: Please consider factoring out this into a separate function.

File: pcbdraw/pcbdraw.py [hunk 6]
```diff
@@ -867,6 +949,24 @@ def main():
...
+    if args.resistor_flip:
+        for r in args.resistor_flip.split(","):
+            if r in tht_resistor_settings:
+                tht_resistor_settings[r]['flip'] = True
+            else:
+                tht_resistor_settings[r] = {'flip': True}
```
Comment (yaqwsx) [inline pcbdraw/pcbdraw.py:968]: Please consider factoring out this into a separate function.

File: pcbdraw/pcbdraw.py [hunk 7]
```diff
@@ -848,6 +928,8 @@ def main():
     parser.add_argument("--dpi", help="DPI for bitmap output", type=int, default=300)
     parser.add_argument("--no-warn-back", action="store_true", help="Don't show warnings about back footprints")
     parser.add_argument("--shrink", type=float, help="Shrink the canvas size to the size of the board. Specify border in millimeters")
+    parser.add_argument("--resistor-values", help="A comma seperated list of what value to set to each resistor for the band colors. For example, \"R1:10k,R2:470\"")
+    parser.add_argument("--resistor-flip", help="A comma seperated list of throughole resistors to flip the bands")
```
Comment (yaqwsx) [inline pcbdraw/pcbdraw.py:932]: This is more-less a philosophical question of whether we should supply these values as arg or from a file. If I want to generate a drawing of a board, I usually want to preserve the setting so I would like to put it into a file. Then I have to invoke it like: `pcbdraw --resistor-values $(cat myfile)`.

I am probably fine with keeping this as it is. I am considering switching to [click](https://click.palletsprojects.com/en/8.0.x/) for the v1 release and then we can address this. Also, as KiKit features schematics parser, we could also supply the values directly from the schematics.

**Review by yaqwsx — CHANGES_REQUESTED** [review]
I walked through the code. It looks good - I just have minor suggestions for improvements.

**Comment by yaqwsx** [comment]
Thank you for your effort! Here are my thoughts on the topics:

- Perhaps there is a better argument input style for the resistor-values: It would be nice to allow for both; direct and file input. However, as I plan major rework of the CLI in the future, is it fine for me to leave this as it is. In the future, we could read the values from schematics.
- The resistor band colors are stored in the board style. Should this be moved to its own style input?: Having it inside the style is OK for me. Once we rework the CLI, we could introduce style inheritance, just like KiKit does with panel configurations. That would solve this issue. 
- If a board style doesn't have a key, it will be copied from the default style. This might be undesired. Thoughts?: As I wrote in the comment in the code - it is OK for me if we also check for extra keys in the style. The reason for that is to catch typos in key names.

I also have one suggestion: Currently, you use opacity to disable the bands. I was thinking of using `display="none"` instead. It is easier to see and we could implement bands that are semi-opaque. Also, I guess `display="none"` might be more efficient on rendering. What are your thoughts?

File: pcbdraw/pcbdraw.py (hunk 1 above)
Comment (Electro707) [reply pcbdraw/pcbdraw.py]: I think instead of finding a resistor value, as any given number for the value can be one, it's better to check if the component supports the color bands as per my recent change:
`if componentElement.find(".//*[@id='{}res_band1']".format(svg_prefix)) is not None:`

File: pcbdraw/pcbdraw.py (hunk 2 above)
Comment (Electro707) [reply pcbdraw/pcbdraw.py]: As above

File: pcbdraw/pcbdraw.py (hunk 3 above)
Comment (Electro707) [reply pcbdraw/pcbdraw.py]: Yeah I agree. I split this into two functions: one that finds the resistor value and tolerance, and another to change the band's color and properties

File: pcbdraw/pcbdraw.py (hunk 6 above)
Comment (Electro707) [reply pcbdraw/pcbdraw.py]: Same as above

File: pcbdraw/pcbdraw.py (hunk 5 above)
Comment (Electro707) [reply pcbdraw/pcbdraw.py]: Moved into a function that creates the entire tht_resistor_settings dictionary.

**Comment by Electro707** [comment]
If a CLI rework is already in the plans, then having it as a direct argument instead of a file is fine.

I do like the idea of changing the `display` property instead of `opacity`, so I changed that in e01e3b0 and the component library.

**Comment by Electro707** [comment]
I pushed bd433e9 which fixes a bug with my previous PR #77. 
The fix

File: pcbdraw/pcbdraw.py (hunk 1 above)
Comment (yaqwsx) [reply pcbdraw/pcbdraw.py]: I like to solution

**Comment by yaqwsx** [comment]
Hi, do you plan to incorporate any changes? If not, I will give the code final proof-reading and testing and we could merge this!

**Comment by Electro707** [comment]
I don't see any more changes that can be added

**Comment by yaqwsx** [comment]
Looks good to me. Could you rebase is to master? Then I will merge this.

**Comment by Electro707** [comment]
Rebased to master, also updated the footprint library to point to master of `pcbdraw-lib`
